sdk/core/core-lro/src/createPoller.ts

```diff
@@ -0,0 +1,162 @@
+   */
+  initState: (config) => ({ status: "running", config } as any),
+  setCanceled: (state) => (state.status = "canceled"),
+  setCanceling: (state) => (state.status = "canceling"),
+  setError: (state, error) => (state.error = error),
+  setResult: (state, result) => (state.result = result),
+  setRunning: (state) => (state.status = "running"),
+  setSucceeded: (state) => (state.status = "succeeded"),
+  setFailed: (state) => (state.status = "failed"),
+
+  getError: (state) => state.error,
+  getResult: (state) => state.result,
+  isCanceled: (state) => state.status === "canceled",
+  isCanceling: (state) => state.status === "canceling",
+  isFailed: (state) => state.status === "failed",
+  isRunning: (state) => state.status === "running",
+  isSucceeded: (state) => state.status === "succeeded",
+});
+
+/**
+ * Creates a poller that can be used to poll a long-running operation.
+ * @param lro - Description of the long-running operation
+ * @param options - options to configure the poller
+ * @returns an initialized poller
+ */
+export async function createPoller<TResult, TState extends OperationState<TResult>>(
+  lro: LongRunningOperation,
+  options?: CreatePollerOptions<TResult, TState>
+): Promise<SimplePollerLike<TState, TResult>> {
+  const {
+    resourceLocationConfig,
+    processResult,
+    updateState,
+    withPollingUrl: withPollingUrlCallback,
+    intervalInMs = POLL_INTERVAL_IN_MS,
+    restoreFrom,
+  } = options || {};
+  const { requestMethod, requestPath } = lro;
+  const stateProxy = createStateProxy<TResult, TState>();
+  const withPollingUrl = withPollingUrlCallback
+    ? (() => {
+        let called = false;
+        return (pollingUrl: string, isUpdated: boolean) => {
+          if (isUpdated) withPollingUrlCallback(pollingUrl);
+          else if (!called) withPollingUrlCallback(pollingUrl);
+          called = true;
+        };
+      })()
+    : undefined;
+  const state: RestorableOperationState<TState> = restoreFrom
+    ? deserializeState(restoreFrom)
+    : await initOperation({
+        lro,
+        stateProxy,
+        requestPath,
+        requestMethod,
+        resourceLocationConfig,
+        processResult,
+        withPollingUrl,
+      });
+  let resultPromise: Promise<TResult> | undefined;
+  let cancelJob: (() => void) | undefined;
+  // Progress handlers
+  type Handler = (state: TState) => void;
+  const handlers = new Map<symbol, Handler>();
+  const handleProgressEvents = async (): Promise<void> => handlers.forEach((h) => h(state));
+
+  let currentPollIntervalInMs = intervalInMs;
+
+  const poller: SimplePollerLike<TState, TResult> = {
+    getOperationState: () => state,
+    getResult: () => state.result,
+    isDone: () => ["succeeded", "failed", "canceled"].includes(state.status),
+    isStopped: () => resultPromise === undefined,
+    stopPolling: () => cancelJob?.(),
+    toString: () =>
+      JSON.stringify({
+        state,
```

Comment:
I am not sure it'll always be safe to serialize operation state as JSON. The operation state might include convenience-layer transforms that make it cyclic, or it could be _extremely_ deep. We should only have to serialize --- in general --- the information that's needed to poll the operation.

The reason `serialize` was a user-provided method in the FR implementation was to account for this possibility and to allow simply serializing the model & operation IDs required to restore the poller, rather than the whole operation state which might include a _really_ deep object.

Reply:
>  The operation state might include convenience-layer transforms that make it cyclic, or it could be extremely deep. 

Is there an existing example of this scenario?

> We should only have to serialize --- in general --- the information that's needed to poll the operation.

I disagree, a client will attach something to the state for a reason, e.g. in TA, the input documents list is attached to the state because it is needed to sort the results.

Reply:
I gave this some more thinking, I guess there is a question about whether all poller states are serializable or not.

Unserializable states are [against our SDK design guidelines](https://azure.github.io/azure-sdk/typescript_design.html#ts-lro-continuation):
>  DO provide a way to instantiate a poller with the serialized state of another poller to begin where it left off, for example by passing the state as a parameter to the same method which started the operation, or by directly instantiating a poller with that state.

 So I think we should require our states to always be JSON serializable. That being said, I would love to take a look at any existing poller with an unserializable state and see if we could make it serializable.

Reply:
> Is there an existing example

In FR, I only serialize the bare minimum of information that is required to restore the poller. I suppose whether or not this is important depends on whether or not you think a completed poller should be restorable, because the full state of a completed poller includes the result, but the serialized version in FR doesn't.

I think it's certainly possible that _in general_ a status monitor could return way more information than we want to serialize into the restoration token, and _in general_ a convenience layer state transform could make the full state cyclic or otherwise unserializable.

> the input documents list is attached to the state because it is needed to sort the results.

The input documents themselves aren't needed, only the input order of their IDs, IIRC. If I call the TA analyze LRO with a couple of megabytes of documents in the input and then serialize the state, do I get a restoration token that is a few megabytes huge?

Reply:
> I think it's certainly possible that in general a status monitor could return way more information than we want to serialize into the restoration token, and in general a convenience layer state transform could make the full state cyclic or otherwise unserializable.

Yeah I was hoping for a concrete example for this situation to discuss further. As I mentioned, generally, an unserializable state is against our design guideline.

> The input documents themselves aren't needed, only the input order of their IDs, IIRC. If I call the TA analyze LRO with a couple of megabytes of documents in the input and then serialize the state, do I get a restoration token that is a few megabytes huge?

Good point, thanks for the suggestion. Currently the service limits the number of input documents to 25 only so the restoration token should contain at worst an array of 25 numbers.

Reply:
I addressed the TA comment in the other PR (https://github.com/Azure/azure-sdk-for-js/pull/22852), so state would look like the following:
```
{"state":{"status":"running","config":{"mode":"OperationLocation","pollingUrl":"https://javatextanalyticstestresources.cognitiveservices.azure.com/language/analyze-text/jobs/<OPERATION ID>?api-version=2022-04-01-preview"},"createdOn":"2022-08-12T12:25:37.000Z","modifiedOn":"2022-08-12T12:25:38.000Z","expiresOn":"2022-08-13T12:25:37.000Z","id":"<OPERATION ID>","actionSucceededCount":0,"actionFailedCount":0,"actionInProgressCount":1,"docIds":["0","1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"]}}
```

Reply:
Are there any outstanding asks in this thread?